### PR TITLE
chore: RFC for extracing Helm charts to a separate repo

### LINKED ISSUE #8447: RFC: Extract helm-charts out of timberio/vector repository
The purpose of this RFC is to propose a migration of the Vector helm charts into a separate repo, decoupling chart releases from application releases and allowing for a more conventional management, testing, and releasing process.

This RFC will solve for the following: 

- [ ] Introducing industry standard tooling to ensure chart reliability
- [ ] Running e2e Kubernetes integration tests for Vector
- [ ] Maintaining up-to-date resources for use with `kustomize` or `kubectl`
- [ ] Updating the charts' `appVersion`'s when new versions of Vector are released

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

Closes #8447

### CODE REVIEW

**Review by binarylogic — APPROVED** [review]
This is an excellent RFC. Thanks for the detailed write up. From the benefits listed here (versioning, improving CI performance, reducing CI scope, easier to maintain), it’s hard to see why we wouldnt do this, but I’m not a k8s expert. I can't poke holes as a result, so I’d like to hear from others.

**Comment by StephenWakely** [comment]
Currently the k8s-e2e-tests will use the charts in `./distribution/helm`. Would the plan be to change that to pull the latest chart from `https://packages.timber.io/helm/latest`?

I guess my only concern is that if we make a change to Vector that also requires a change to the Helm charts, the tests would fail since the latest Helm chart can't be pushed until the changed version of Vector has also been released. 

Is my understanding correct here, and what would be the solution to this?

**Comment by spencergilbert** [comment]
> Currently the k8s-e2e-tests will use the charts in `./distribution/helm`. Would the plan be to change that to pull the latest chart from `https://packages.timber.io/helm/latest`?
> 
> I guess my only concern is that if we make a change to Vector that also requires a change to the Helm charts, the tests would fail since the latest Helm chart can't be pushed until the changed version of Vector has also been released.
> 
> Is my understanding correct here, and what would be the solution to this?

There should be no cases where we couldn't update the helm charts prior to updating the application and changes should be forward/backward compatible for that reason as well.

Do you have an example where this would actually be an issue?

Further more, we are over coupling the Kubernetes integration tests with the helm chart as it stands today and this move would improve that and we should aim to reduce the dependencies of the `kubernetes_logs` (kubernetes integration) on the helm charts directly.

**Comment by StephenWakely** [comment]
> Do you have an example where this would actually be an issue?

No, I haven't. Was just grasping for potential problems..

**Review by StephenWakely — APPROVED** [review]
Nice work! It sounds like this is a much better approach.

**Comment by spencergilbert** [comment]
> > Do you have an example where this would actually be an issue?
> 
> No, I haven't. Was just grasping for potential problems..

👍 I can't think of anything myself, but also can't think of a case where an (at least initially) opt-in config on the helm side would block the helm ci and not unblock the application

**Comment by spencergilbert** [comment]
Based on discussions of Q3 plans with @binarylogic, @zsherman, and @jszwedko - I'm going to update this RFC to propose moving our charts directly to the `DataDog/helm-charts` repository

**Comment by spencergilbert** [comment]
I've started working on the migration PR here: https://github.com/spencergilbert/dd-helm-charts/pull/1

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (JeanMertz) [inline rfcs/2021-07-23-8447-extract-helm-charts.md:112]: A few additional questions:

1. The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?
2. Do we need to update existing contributing docs
3. Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?
4. Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

**Review by JeanMertz — APPROVED** [review]
I left a few remarks, but overall this looks good to me 👍

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:47]: 😞

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:112]: > 1. The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?

Ah, good call - I didn't actually check, @binarylogic do you have any thoughts here. I'll check with the container-integrations team in the meantime.

> 2. Do we need to update existing contributing docs

I believed we would assume Datadog's [existing contributing docs](https://github.com/DataDog/helm-charts/blob/master/CONTRIBUTING.md) but could call that out specifically.

> 3. Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?

No CLA or DCO sort of thing needed.

> 4. Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

We'll get access to the repository and get codeowner status on our charts 👍

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:55]: I'll be more specific here, but it was the "general" concept of our CI jobs for helm charts - lint, kubeval, and release. We won't use the exact same tooling, but the intent doesn't change.

More generally, we should not expect to lose any CI functionality with this migration.

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:18]: Will update - it's worded strangely. The intent was "I am not proposing any changes to the current raw and kustomize resources", though there are changes I may propose in the future 😉

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:30]: 👍 added [this link](https://app.datadoghq.com/metric/explorer?from_ts=1627251689494&to_ts=1627445043208&live=false&tile_size=l&exp_metric=gh.actions.workflow_job.execution_secs.99percentile&exp_scope=conclusion%3Asuccess%2Cworkflow%3Ak8s_e2e_suite&exp_group=workflow&exp_agg=max&exp_row_type=metric#workflow:test_suite) for our current tests

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (jszwedko) [inline rfcs/2021-07-23-8447-extract-helm-charts.md:60]: These seem like details we can figure out later, but for the purposes of RFCs we typically prefer to pick a specific path.

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (jszwedko) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:33]: True, it does seem like we'll be involving a separate release cycle here including its own announcements. It isn't 100% clear how we would have those announcements on the website (or if we need to?).

**Review by jszwedko — APPROVED** [review]
Thanks for this detailed write-up! I think the proposal makes sense and is in-line with how Datadog manages the charts for the datadog-agent.

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:33]: I'd consider most of the changes made to the helm charts to not be worth announcing. I'd continue to defer to "how datadog does it"

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:60]: Fair, the differences are fairly slight and I'd implement whichever was least effort. I can do the research ahead of time if need be.

**Review by StephenWakely — APPROVED** [review]

File: rfcs/2021-07-23-8447-extract-helm-charts.md
Comment (jszwedko) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:112]: Just following up that we decided on keeping a separate repo in the `timberio` org for now which I think obviates some of these questions:

> * The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?

We can keep our MPL 2.0 license

> * Do we need to update existing contributing docs

It does seem like we'll need to do this to point users to in in our `CONTRIBUTING.md`

> * Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?

We can avoid this for now.

> * Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

Same, we'll use our current processes.

> - Do we need to automate the cross-repository management or just document the steps with the initial work?

Could you expand a bit on what this means? I feel like I'm lacking some context.

> - Do we need to migrate old chart releases to the `DataDog/helm-charts` repository, or direct users to the old repository for old releases?

Is this still an issue if we keep it in the `timberio` org for now? Would it be the same helm chart repository?

Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:112]: > > * Do we need to automate the cross-repository management or just document the steps with the initial work?
> 
> Could you expand a bit on what this means? I feel like I'm lacking some context.

Yep! When we cut a new Vector release. we'll want to update the Helm chart to use the image and in the opposite direction, if we make material changes to the charts that we need to regenerate the `distribution/kubernetes` contents. Does that make sense? It's probably possible to automate these, but otherwise can be scripts/make targets that we run when needed.

> > * Do we need to migrate old chart releases to the `DataDog/helm-charts` repository, or direct users to the old repository for old releases?
> 
> Is this still an issue if we keep it in the `timberio` org for now? Would it be the same helm chart repository?

That's a good question - we _could_ update our custom release script and continue using the existing s3 repo. If we wanted to adopt [chart-releaser](https://github.com/helm/chart-releaser) we'd end up with a new repo url and would need to consider migrating.

Comment (jszwedko) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:112]: > Yep! When we cut a new Vector release. we'll want to update the Helm chart to use the image and in the opposite direction, if we make material changes to the charts that we need to regenerate the `distribution/kubernetes` contents. Does that make sense? It's probably possible to automate these, but otherwise can be scripts/make targets that we run when needed.
> 
> > >

Ah, I see. I think we can start with just documenting the procedure and then automating it if it becomes a pain.



> That's a good question - we _could_ update our custom release script and continue using the existing s3 repo. If we wanted to adopt [chart-releaser](https://github.com/helm/chart-releaser) we'd end up with a new repo url and would need to consider migrating.

As long as we don't have to have everyone migrate again if/when we rename the `timberio` org then I think that is ok to ask them to migrate once as part of this cut over. Regarding old chart releases, would it be difficult to copy them?

Comment (spencergilbert) [reply rfcs/2021-07-23-8447-extract-helm-charts.md:112]: > As long as we don't have to have everyone migrate again if/when we rename the timberio org then I think that is ok to ask them to migrate once as part of this cut over. Regarding old chart releases, would it be difficult to copy them?

If we re-released old charts we could handle user migrations with redirects... We could do something _weird_ by doing tags/releases and have the git history still... or manually publish them but not have them in the git history

**Pull request merged**
